lib/pipelineTemplateVersionFactory.js

```diff
@@ -0,0 +1,153 @@
+                        namespace: config.namespace,
+                        name: config.name,
+                        maintainer: config.maintainer,
+                        createTime,
+                        updateTime: createTime
+                    });
+                }
+
+                return pipelineTemplate;
+            })
+            .then(pipelineTemplate => {
+                const isExactVersion = EXACT_VERSION_REGEX.exec(config.version);
+
+                if (isExactVersion) {
+                    return this.get({
+                        templateId: pipelineTemplate.id,
+                        version: config.version
+                    }).then(versionedConfig => {
+                        return { pipelineTemplate, versionedConfig };
+                    });
+                }
+
+                return { pipelineTemplate, versionedConfig: null, isExactVersion };
+            })
+            .then(result => {
+                const { pipelineTemplate, versionedConfig, isExactVersion } = result;
+                const [, configMajor, configMinor] = VERSION_REGEX.exec(config.version);
+
+                let newVersion;
+
+                if (!versionedConfig) {
+                    if (!isExactVersion) {
+                        newVersion = config.version;
+                    } else if (!pipelineTemplate.latestVersion) {
+                        newVersion = configMinor ? `${configMajor}${configMinor}.0` : `${configMajor}.0.0`;
+                    } else {
+                        const [, globalMajor, globalMinor, globalPatch] = VERSION_REGEX.exec(
+                            pipelineTemplate.latestVersion
+                        );
+                        const patch = parseInt(globalPatch.slice(1), 10) + 1;
+
+                        newVersion = `${globalMajor}${globalMinor}.${patch}`;
+                    }
+                }
+
+                return super
+                    .create({
+                        templateId: pipelineTemplate.id,
+                        description: config.description,
+                        config: config.config,
+                        createTime,
+                        version: newVersion
+                    })
+                    .then(pipelineTemplateVersion => {
+                        return { pipelineTemplateVersion, pipelineTemplate };
+                    });
+            })
+            .then(result => {
+                const { pipelineTemplate, pipelineTemplateVersion } = result;
+                const [, latestMajor, latestMinor] = VERSION_REGEX.exec(pipelineTemplate.latestVersion);
+                const [, major, minor] = VERSION_REGEX.exec(pipelineTemplateVersion.version);
+
+                if (major > latestMajor || (major === latestMajor && minor >= latestMinor)) {
+                    pipelineTemplate.latestVersion = pipelineTemplateVersion.version;
+                    pipelineTemplate.updateTime = new Date().toISOString();
+                }
```

Comment:
We should not update the latest version when `major` and `minor` matches unless the `patch` of newly created template is greater than the current latest `patch` 
```suggestion
                const [, latestMajor, latestMinor, latestPatch] = VERSION_REGEX.exec(pipelineTemplate.latestVersion);
                const [, major, minor, latestMinor] = VERSION_REGEX.exec(pipelineTemplateVersion.version);

                if (major > latestMajor 
                  || (major === latestMajor && minor > latestMinor)
                  || (major === latestMajor && minor === latestMinor && patch > latestPatch)) {
                    pipelineTemplate.latestVersion = pipelineTemplateVersion.version;
                    pipelineTemplate.updateTime = new Date().toISOString();
                }
```

Reply:
Yes, I will revisit this